### PR TITLE
Bug 1842039 - Add bread crumbs for IllegalArgumentException when navigating

### PR DESCRIPTION
**Pull request by Amejia481**
Tried to add as many scenarios from [crash-stats](https://crash-stats.mozilla.org/signature/?signature=java.lang.IllegalArgumentException%3A%20at%20androidx.navigation.NavController.navigate%28NavController.kt%3A23%29&date=%3E%3D2023-08-05T01%3A03%3A00.000Z&date=%3C2023-08-12T01%3A03%3A00.000Z&_columns=date&_columns=product&_columns=version&_columns=build_id&_columns=platform&_columns=reason&_columns=address&_columns=install_time&_columns=startup_crash&_sort=-date&page=1#reports) (Until page 2). 

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1842039

### CODE REVIEW

**Comment by Amejia481** [comment]
@jonalmeida I updated the PR with changes we discussed yesterday :)

**Review by jonalmeida — APPROVED** [review]
Looks good, just some nits!

Commit message nit: We are not addressing this crash with our change, so maybe something like "Add breadcrumbs for IllegalArgumentException navigation"?

Could you also file a follow-up linked bug to analyze the breadcrumbs once this change lands?

File: android-components/components/lib/crash/docs/metrics.md
[the remark this answers is not in the document]
Comment (Amejia481) [reply android-components/components/lib/crash/docs/metrics.md:1]: Thank for catching it.
It was unintentional  😅!

**Comment by Amejia481** [comment]
> Looks good, just some nits!
> 
> Commit message nit: We are not addressing this crash with our change, so maybe something like "Add breadcrumbs for IllegalArgumentException navigation"?
> 
> Could you also file a follow-up linked bug to analyze the breadcrumbs once this change lands?

Sure!

**Comment by Amejia481** [comment]
Addressed the comments and file for further investigation https://bugzilla.mozilla.org/show_bug.cgi?id=1848844

File: fenix/app/src/main/java/org/mozilla/fenix/ext/NavController.kt [hunk 1]
```diff
@@ -36,6 +39,29 @@ fun NavController.navigateSafe(
...
+ */
+fun NavController.navigateWithBreadcrumb(
+    directions: NavDirections,
+    navigateFrom: String,
+    navigateTo: String,
+    crashReporter: CrashReporter,
+) {
+    try {
+        this.navigate(directions)
+    } catch (e: IllegalArgumentException) {
```
Comment (jonalmeida) [inline fenix/app/src/main/java/org/mozilla/fenix/ext/NavController.kt:54]: Ack, @Amejia481 I realized that this is a catch so the crash exception won't end up in Sentry. We probably need to add the breadcrumb, and then let the crash happen.

Comment (Amejia481) [reply fenix/app/src/main/java/org/mozilla/fenix/ext/NavController.kt:54]: Yeah you are right, I think we missed the submit the exception.

**Comment by Amejia481** [comment]
@Mergifyio backport releases_v117

**Comment by Amejia481** [comment]
@Mergifyio backport releases_v116

**Pull request merged**